File: shared/host-numerics/examples/tutorial.py
# Copyright Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT

"""Executable walkthrough of the public Python API.

The numbered tests are intentionally small and ordered like a tutorial. CTest
runs this file so the examples stay synchronized with the extension module.
"""

import unittest

import numpy as np

import roc_host_numerics as hn


def real_recipe(component, seed):
    return hn.GenerationRecipe.real_only(
        component,
        hn.GenerationRecipeSettings(
            seed=seed,
            index_order=hn.IndexOrder.LastDimensionFastest,
        ),
    )


class HostNumericsTutorial(unittest.TestCase):
    def test_01_create_and_convert_tensors(self):
        values = np.asarray([[1.0, 2.0], [3.0, 4.0]], dtype=np.float32)
        tensor = hn.from_numpy(values)
        self.assertEqual(tensor.type, hn.ScalarType.Float32)
        self.assertEqual(tensor.shape, [2, 2])
        np.testing.assert_array_equal(hn.to_numpy(tensor), values)

        fp4 = hn.from_numpy(
            np.asarray([0.0, 0.5, 1.0, 6.0], dtype=np.float32),
            hn.ScalarType.Float4E2M1,
        )
        self.assertEqual(len(fp4.storage), 2)
        np.testing.assert_array_equal(
            hn.to_numpy(fp4), np.asarray([0.0, 0.5, 1.0, 6.0], dtype=np.float32)
        )

    def test_02_generate_reproducible_tensors(self):
        real = real_recipe(
            hn.GenerationRecipe.uniform_real(
                hn.UniformRealGenerationParameters(lower=-1.0, upper=1.0)
            ),
            seed=17,
        )
        first = hn.generate_tensor(hn.ScalarType.Float32, [8], real)
        replay = hn.generate_tensor(hn.ScalarType.Float32, [8], real)
        self.assertEqual(first.storage, replay.storage)

        integers = hn.generate_tensor(
            hn.ScalarType.Int32,
            [8],
            real_recipe(
                hn.GenerationRecipe.uniform_integer(
                    hn.UniformIntegerGenerationParameters(lower=1, upper=4)
                ),
                seed=27,
            ),
        )
        self.assertTrue(np.any(hn.to_numpy(integers) != 0))

        complex_values = hn.generate_tensor(
            hn.ScalarType.ComplexFloat32,
            [8],
            hn.GenerationRecipe.cartesian(
                hn.GenerationRecipe.uniform_integer(
                    hn.UniformIntegerGenerationParameters(lower=1, upper=4)
                ),
                hn.GenerationRecipe.uniform_integer(
                    hn.UniformIntegerGenerationParameters(lower=-4, upper=-1)
                ),
                hn.GenerationRecipeSettings(seed=37),
            ),
        )
        self.assertTrue(np.all(hn.to_numpy(complex_values).imag != 0))

    def test_03_compose_broadcast_tensor_operations(self):
        column = hn.from_numpy(np.asarray([[1.0], [2.0]], dtype=np.float32))
        row = hn.from_numpy(np.asarray([[10.0, 20.0, 30.0]], dtype=np.float32))
        result = 2.0 * column + row
        np.testing.assert_array_equal(
            hn.to_numpy(result),
            np.asarray([[12.0, 22.0, 32.0], [14.0, 24.0, 34.0]], dtype=np.float32),
        )
        np.testing.assert_array_equal(
            hn.to_numpy(hn.relu(hn.clip(result, -20.0, 20.0))),
            np.asarray([[12.0, 20.0, 20.0], [14.0, 20.0, 20.0]], dtype=np.float32),
        )

    def test_04_multiply_matrices(self):
        a_values = np.asarray([[1, 2, 3], [4, 5, 6]], dtype=np.float32)
        b_values = np.asarray([[7, 8], [9, 10], [11, 12]], dtype=np.float32)
        product = hn.matmul(hn.from_numpy(a_values), hn.from_numpy(b_values))
        np.testing.assert_array_equal(hn.to_numpy(product), a_values @ b_values)

        integer_product = hn.matmul(
            hn.from_numpy(np.asarray([[2, 3]], dtype=np.int8)),
            hn.from_numpy(np.asarray([[4], [5]], dtype=np.int8)),
            output_type=hn.ScalarType.Int32,
            accumulator_type=hn.ScalarType.Int32,
        )
        np.testing.assert_array_equal(
            hn.to_numpy(integer_product), np.asarray([[23]], dtype=np.int32)
        )

        complex_a = np.asarray([[1 + 1j, 2 - 1j]], dtype=np.complex64)
        complex_b = np.asarray([[3 + 0j], [1 + 2j]], dtype=np.complex64)
        complex_product = hn.matmul(
            hn.from_numpy(complex_a),
            hn.from_numpy(complex_b),
            output_type=hn.ScalarType.ComplexFloat32,
            accumulator_type=hn.ScalarType.ComplexFloat32,
            conjugate_a=True,
        )
        np.testing.assert_array_equal(
            hn.to_numpy(complex_product), np.conjugate(complex_a) @ complex_b
        )

        empty_product = hn.matmul(
            hn.from_numpy(np.empty((2, 0), dtype=np.float32)),
            hn.from_numpy(np.empty((0, 3), dtype=np.float32)),
        )
        np.testing.assert_array_equal(
            hn.to_numpy(empty_product), np.zeros((2, 3), dtype=np.float32)
        )

    def test_05_generate_mxfp4_data_and_scales(self):
        recipe = real_recipe(
            hn.GenerationRecipe.uniform_real(
                hn.UniformRealGenerationParameters(lower=-1.0, upper=1.0)
            ),
            seed=47,
        )
        data = hn.MxDataGeneration.preserve_range(
            recipe, hn.MxRepresentedValueRange(lower=-1.0, upper=1.0)
        )
        options = hn.MxGenerationOptions()
        options.data_type = hn.ScalarType.Float4E2M1
        options.scale_type = hn.ScalarType.E8M0
        options.leading_dimension = 32
        options.block_axis = 0
        options.block_size = 32
        mx = hn.generate_mx([32, 2], data, options)

        self.assertEqual(mx.data.type, hn.ScalarType.Float4E2M1)
        self.assertEqual(mx.scales.shape, [2, 1])
        self.assertEqual(mx.scale_indices.shape, mx.data.shape)
        self.assertEqual(mx.reference.shape, mx.data.shape)

    def test_06_use_reductions_normalization_and_sparsity(self):
        values = hn.from_numpy(np.asarray([[1.0, 2.0], [3.0, 4.0]], dtype=np.float32))
        sums = hn.reference_sum(
            values,
            output_type=hn.ScalarType.Float32,
            accumulator_type=hn.ScalarType.Float32,
            axes=[1],
        )
        np.testing.assert_array_equal(hn.to_numpy(sums), np.asarray([3.0, 7.0]))

        probabilities = hn.reference_softmax(values, axis=1)
        np.testing.assert_allclose(
            np.sum(hn.to_numpy(probabilities), axis=1), np.ones(2), rtol=1e-6
        )

        normalized = hn.reference_layer_norm(
            values,
            output_type=hn.ScalarType.Float32,
            statistics_type=hn.ScalarType.Float32,
            axis=1,
        )
        self.assertEqual(normalized.mean.shape, [2])

        pattern = hn.StructuredSparsityPattern()
        pattern.axis = 1
        pattern.fixed_positions = [0, 2]
        sparse = hn.apply_structured_sparsity(
            hn.from_numpy(np.asarray([[1.0, 2.0, 3.0, 4.0]], dtype=np.float32)),
            pattern,
            emit_two_of_four_metadata=True,
        )
        self.assertEqual(hn.to_numpy(sparse.pruned)[0, 1], 0.0)
        self.assertIsNotNone(sparse.retained_indices)
        self.assertIsNotNone(sparse.two_of_four_metadata)

    def test_07_inspect_a_comparison_failure(self):
        expected = hn.from_numpy(np.asarray([1.0, 2.0, 3.0], dtype=np.float32))
        observed = hn.from_numpy(np.asarray([1.0, 9.0, 3.0], dtype=np.float32))
        report = hn.compare(observed, expected)
        self.assertFalse(report.passed)
        self.assertEqual(report.mismatches, 1)
        self.assertTrue(
            str(report).startswith("comparison failed: 1 of 3 compared elements mismatched")
        )
        self.assertIn("index 1 [1]", str(report))


if __name__ == "__main__":
    unittest.main()

File: shared/host-numerics/examples/tutorial.cpp
// Copyright Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT

// This executable is both a tutorial and a CTest contract test. Keep the
// numbered examples in a readable order and prefer small, inspectable values.

#include <algorithm>
#include <array>
#include <complex>
#include <iostream>
#include <roc/host_numerics/comparison.hpp>
#include <roc/host_numerics/gemm.hpp>
#include <roc/host_numerics/generation.hpp>
#include <roc/host_numerics/layer_norm.hpp>
#include <roc/host_numerics/mx.hpp>
#include <roc/host_numerics/reduction.hpp>
#include <roc/host_numerics/softmax.hpp>
#include <roc/host_numerics/structured_sparsity.hpp>
#include <roc/host_numerics/tensor_operations.hpp>
#include <span>
#include <stdexcept>
#include <string>

namespace {
using namespace roc::host_numerics;

void check(bool condition, const char* message) {
    if (!condition) throw std::runtime_error(message);
}

// Step 1: create tensors from ordinary values. ScalarType records the encoded
// storage format, while Shape and Layout describe logical indexing.
void tensorBasics() {
    const Tensor values =
        Tensor::copyNativeValues<float>(Shape{2, 2}, std::array<float, 4>{1.0f, 2.0f, 3.0f, 4.0f});
    check(values.type() == ScalarType::Float32 && values.shape() == Shape{2, 2},
          "Float32 tensor metadata mismatch.");

    const Tensor scalar = Tensor::scalar(ScalarType::Float16, 1.5f);
    check(scalar.shape() == Shape{} && scalar.item<float>() == 1.5f,
          "Rank-zero tensor conversion mismatch.");

    const std::array<float, 4> fp4Values{0.0f, 0.5f, 1.0f, 6.0f};
    const Tensor fp4 = Tensor::copyValuesWithConversion(ScalarType::Float4E2M1, Shape{4},
                                                        std::span<const float>(fp4Values));
    check(fp4.rawEncodedBackingStorage().size() == 2 && fp4.loadAs<float>({3}) == 6.0f,
          "Packed FP4 tensor mismatch.");
}

// Step 2: generate reproducible real, integer, and complex tensors. A caller
// supplies one seed per logical operand; replaying a seed reproduces its bytes.
void deterministicGeneration() {
    const GenerationRecipe realRecipe = GenerationRecipe::realOnly(
        GenerationRecipe::uniformReal({.lower = -1.0, .upper = 1.0}), {.seed = 17});
    const Tensor real = generate(ScalarType::Float32, Shape{8}, realRecipe);
    const Tensor replay = generate(ScalarType::Float32, Shape{8}, realRecipe);
    check(std::equal(real.rawEncodedBackingStorage().begin(), real.rawEncodedBackingStorage().end(),
                     replay.rawEncodedBackingStorage().begin(),
                     replay.rawEncodedBackingStorage().end()),
          "Equal generation seeds produced different storage.");

    const Tensor integers =
        generate(ScalarType::Int32, Shape{8},
                 GenerationRecipe::realOnly(
                     GenerationRecipe::uniformInteger({.lower = 1, .upper = 4}), {.seed = 27}));
    bool anyNonzero = false;
    for (size_t index = 0; index < integers.elementCount(); ++index)
        anyNonzero |= integers.loadAs<int32_t>({index}) != 0;
    check(anyNonzero, "Integer generation unexpectedly produced only zeros.");

    const Tensor complex =
        generate(ScalarType::ComplexFloat32, Shape{8},
                 GenerationRecipe::cartesian(
                     GenerationRecipe::uniformInteger({.lower = 1, .upper = 4}),
                     GenerationRecipe::uniformInteger({.lower = -4, .upper = -1}), {.seed = 37}));
    check(complex.loadAs<std::complex<float>>({0}).imag() != 0.0f,
          "Complex generation omitted its imaginary component.");
}

// Step 3: compose broadcast elementwise operations. Operators are shorthand
// for the same-type defaults; named calls expose compute/output policy.
void tensorArithmetic() {
    const Tensor column =
        Tensor::copyNativeValues<float>(Shape{2, 1}, std::array<float, 2>{1.0f, 2.0f});
    const Tensor row =
        Tensor::copyNativeValues<float>(Shape{1, 3}, std::array<float, 3>{10.0f, 20.0f, 30.0f});
    const Tensor result = 2.0f * column + row;
    check(result.shape() == Shape{2, 3} && result.loadAs<float>({0, 0}) == 12.0f &&
              result.loadAs<float>({1, 2}) == 34.0f,
          "Broadcast tensor arithmetic mismatch.");
    check(relu(clip(result, -20.0, 20.0)).loadAs<float>({1, 2}) == 20.0f,
          "Named activation mismatch.");
}

// Step 4: matrix multiplication keeps storage and arithmetic types explicit.
void matrixMultiplication() {
    const Tensor a =
        Tensor::copyNativeValues<float>(Shape{2, 3}, std::array<float, 6>{1, 2, 3, 4, 5, 6});
    const Tensor b =
        Tensor::copyNativeValues<float>(Shape{3, 2}, std::array<float, 6>{7, 8, 9, 10, 11, 12});
    const Tensor product = matmul(a, b, ScalarType::Float32);
    check(compare(product, Tensor::copyNativeValues<float>(Shape{2, 2},
                                                           std::array<float, 4>{58, 64, 139, 154}))
              .passed(),
          "Float32 matmul mismatch.");

    MatmulOptions integerOptions(ScalarType::Int32);
    const Tensor integerProduct =
        matmul(Tensor::copyNativeValues<int8_t>(Shape{1, 2}, std::array<int8_t, 2>{2, 3}),
               Tensor::copyNativeValues<int8_t>(Shape{2, 1}, std::array<int8_t, 2>{4, 5}),
               ScalarType::Int32, integerOptions);
    check(integerProduct.loadAs<int32_t>({0, 0}) == 23, "Int8-to-Int32 matmul mismatch.");

    MatmulOptions complexOptions(ScalarType::ComplexFloat32);
    complexOptions.conjugateA = true;
    const Tensor complexProduct =
        matmul(Tensor::copyNativeValues<std::complex<float>>(
                   Shape{1, 2}, std::array{std::complex<float>{1, 1}, std::complex<float>{2, -1}}),
               Tensor::copyNativeValues<std::complex<float>>(
                   Shape{2, 1}, std::array{std::complex<float>{3, 0}, std::complex<float>{1, 2}}),
               ScalarType::ComplexFloat32, complexOptions);
    check(complexProduct.loadAs<std::complex<float>>({0, 0}) == std::complex<float>(3, 2),
          "Conjugated complex matmul mismatch.");

    const Tensor emptyProduct =
        matmul(Tensor(ScalarType::Float32, Shape{2, 0}), Tensor(ScalarType::Float32, Shape{0, 3}),
               ScalarType::Float32);
    check(emptyProduct.shape() == Shape{2, 3} && emptyProduct.loadAs<float>({1, 2}) == 0.0f,
          "Zero-reduction matmul is not the additive identity.");
}

// Step 5: MX generation returns packed data, natural block scales, a scale map,
// and decoded reference values together.
void blockScaledMxGeneration() {
    const GenerationRecipe recipe = GenerationRecipe::realOnly(
        GenerationRecipe::uniformReal({.lower = -1.0, .upper = 1.0}), {.seed = 47});
    const MxDataGeneration data =
        MxDataGeneration::preserveRange(recipe, {.lower = -1.0, .upper = 1.0});
    MxGenerationOptions options;
    options.dataType = ScalarType::Float4E2M1;
    options.scaleType = ScalarType::E8M0;
    options.leadingDimension = 32;
    options.blockAxis = 0;
    options.blockSize = 32;
    const MxTensor mx = generateMx(Shape{32, 2}, data, options);
    check(mx.data.type() == ScalarType::Float4E2M1 && mx.scales.shape() == Shape{2, 1} &&
              mx.scaleIndices.shape() == mx.data.shape() && mx.reference.shape() == mx.data.shape(),
          "MXFP4 generation result mismatch.");
}

// Step 6: higher-level references compose the same Tensor model.
void higherLevelOperations() {
    const Tensor input =
        Tensor::copyNativeValues<float>(Shape{2, 2}, std::array<float, 4>{1.0f, 2.0f, 3.0f, 4.0f});
    const Tensor sums = referenceSum(input, {1}, ScalarType::Float32, ScalarType::Float32);
    check(sums.shape() == Shape{2} && sums.loadAs<float>({0}) == 3.0f &&
              sums.loadAs<float>({1}) == 7.0f,
          "Reduction mismatch.");

    const Tensor probabilities =
        referenceSoftmax(input, 1, ScalarType::Float32, ScalarType::Float32);
    check(std::abs(probabilities.loadAs<float>({0, 0}) + probabilities.loadAs<float>({0, 1}) -
                   1.0f) < 1e-6f,
          "Softmax row does not sum to one.");

    const LayerNormOutputs normalized =
        referenceLayerNorm(input, {.output = ScalarType::Float32, .mean = ScalarType::Float32});
    check(normalized.mean && normalized.mean->shape() == Shape{2},
          "LayerNorm statistics mismatch.");

    StructuredSparsityPattern pattern;
    pattern.axis = 1;
    pattern.fixedPositions = {0, 2};
    const StructuredSparseTensor sparse = applyStructuredSparsity(
        Tensor::copyNativeValues<float>(Shape{1, 4}, std::array<float, 4>{1, 2, 3, 4}), pattern,
        {.retainedIndices = true, .twoOfFourMetadata = true});
    check(sparse.pruned.loadAs<float>({0, 1}) == 0.0f && sparse.retainedIndices &&
              sparse.twoOfFourMetadata,
          "Structured sparsity outputs mismatch.");
}

// Step 7: comparison returns structured evidence and a reusable diagnostic.
void comparisonDiagnostics() {
    const Tensor expected =
        Tensor::copyNativeValues<float>(Shape{3}, std::array<float, 3>{1, 2, 3});
    const Tensor observed =
        Tensor::copyNativeValues<float>(Shape{3}, std::array<float, 3>{1, 9, 3});
    const ComparisonReport report = compare(observed, expected);
    const std::string message = formatComparisonReport(report);
    check(!report.passed() &&
              message.starts_with("comparison failed: 1 of 3 compared elements mismatched") &&
              message.find("index 1 [1]") != std::string::npos,
          "Comparison failure diagnostic mismatch.");
}
}  // namespace

int main() {
    try {
        tensorBasics();
        deterministicGeneration();
        tensorArithmetic();
        matrixMultiplication();
        blockScaledMxGeneration();
        higherLevelOperations();
        comparisonDiagnostics();
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "host-numerics tutorial failed: " << error.what() << '\n';
        return 1;
    }
}
